### ISSUE TITLE
False positive for question_mark on borrowed content

### ISSUE THREAD

**Issue by fuerstenau** [rust-lang/rust-clippy#3560]
# Version
> $ cargo clippy -V
> 0.0.212

# Code & Confusion
Hopefully sufficient extract:

```rust
fn next(&mut self) -> Option<Self::Item> {
	// Fail quickly when at the end
	if self.current_row.is_none() {
		return None;
	};
	// More code could follow here.
	// Hier hingegen könnte ihre Werbung stehen.
}
```

When linting this, Clippy remarks:
> this block may be rewritten with the `?` operator
> […]
> help: replace_it_with: `self.current_row?;`

Hence, we obtain:

```rust
fn next(&mut self) -> Option<Self::Item> {
	// Fail quickly when at the end
	self.current_row?;
	// More code (but only if we want).
	// Advertising with us is cheap!
	// Only one quatloo per comment that roughly three people will read.
	// (When we include me in the count. Twice.)
}
```

Then, the compiler complains:
> error[E0507]: cannot move out of borrowed content

# Speculation
Until someone tells me otherwise, I'll assume that the `?` operator is not compatible with borrowed content and Clippy fails to check for this.

**Comment by oli-obk** [comment]
Hmm... We have two ways to go on from here:

1. suggest `self.current_row.as_ref()?`
2. do not suggest anything

I'm leaning towards 1. just for the fun of it.

Anyway, to detect this case we need to check the type for [`moves_by_default`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.moves_by_default) and if it does, do one of the two solutions above.

**Comment by fuerstenau** [comment]
Winter break is coming up and that seems fun. Since this is apparently a “good first issue” and I'm here anyway, I'd like to take a stab at it.

**Comment by oli-obk** [comment]
Great!

You can get the type by copying https://github.com/rust-lang/rust-clippy/blob/379c934f3f0c1266e2a4112a4babe7ec0a6368ce/clippy_lints/src/question_mark.rs#L122

The arguments to `moves_by_default` can usually be obtained from the `LateContext` (e.g. `param_env`). Other than that it's just some moving around of the code in the file linked above

**Comment by oli-obk** [comment]
> In the last case, however, using self.opt? instead of self.opt.as_ref()? is actually possible. 

That's very hard to figure out in general. I think it's fine. We can revisit if we actually encounter real world examples.

> Apart from the fact that CONTRIBUTING.md could use some work

Specific complaints are very welcome. It's hard to improve when we don't know the stumbling stones.

**Comment by fuerstenau** [comment]
See PR #3561. That should solve this particular issue.

> Specific complaints are very welcome. It's hard to improve when we don't know the stumbling stones.
I'll try to write something on the matter. If there were only a few well-visible stumbling stones, I would have already pointed them out.

**Comment by ThomasdenH** [comment]
Since the PR has been merged, should this be closed? I think [this](https://github.com/rust-lang/rust-clippy/issues/3560#issuecomment-448204308) deserves its own issue.
